plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/jobs/DataSourceMonitorJob.java

```diff
@@ -77,19 +84,82 @@ protected IStatus run(DBRProgressMonitor monitor) {
         if (platform.isShuttingDown()) {
             return Status.OK_STATUS;
         }
-        if (lastPingTime > 0 && System.currentTimeMillis() - lastPingTime > SYSTEM_SUSPEND_INTERVAL) {
-            log.debug("System suspend detected! Reinitialize all remote connections.");
+        boolean invalidateOnSleep = DBWorkbench.getPlatform().getPreferenceStore().getBoolean(ModelPreferences.CONNECTION_CLOSE_ON_SLEEP);
+        boolean wasSleeping = isSleeping;
+        isSleeping = OperationSystemState.isInSleepMode();
+
+        if (!isSleeping) {
+            if (invalidateOnSleep && lastPingTime > 0 && System.currentTimeMillis() - lastPingTime > SYSTEM_SUSPEND_INTERVAL) {
+                if (INVALIDATE_AFTER_SLEEP) {
+                    invalidateSleptConnections(monitor);
+                }
+            }
+
+            doJob();
+        } else if (!wasSleeping) {
+            // Sleep mode triggered
+            if (invalidateOnSleep) {
+                // Disconnect all datasources
+                closeAllConnections(monitor);
+            }
         }
         lastPingTime = System.currentTimeMillis();
-
-        doJob();
-
         if (!platform.isShuttingDown()) {
             scheduleMonitor();
         }
         return Status.OK_STATUS;
     }
 
+    private void closeAllConnections(DBRProgressMonitor monitor) {
+        log.debug("System suspend detected. Close all remote connections.");
+        final DBPWorkspace workspace = platform.getWorkspace();
+        for (DBPProject project : new ArrayList<>(workspace.getProjects())) {
+            if (project.isOpen() && project.isRegistryLoaded()) {
+                for (DBPDataSourceContainer ds : project.getDataSourceRegistry().getDataSources()) {
+                    if (ds.isConnected() && !ds.getDriver().isEmbedded()) {
+                        try {
+                            ds.disconnect(monitor);
+                        } catch (Exception e) {
+                            log.debug("Error closing connection in sleep mode");
```

Comment:
Would be nice to log which data source failed to disconnect along with the original exception or its message

Reply:
Agreed